### PR TITLE
ztest: error out when building tests with no compiler optimizations

### PR DESCRIPTION
**Pull request by fabiobaltieri**
Follow up from the discussion at the release meeting, feel free to suggest better wording.

--- 

Many tests are known to fail when built and no compiler optimizations. Add a precompiler check to error out when building a ztest based test with no optimization, ask not file issues about it but also adds an opt-out option to bypass the error for tests are actually designed to work in this setup.

### CODE REVIEW

**Review by aaronemassey — DISMISSED** [review]
LGTM as far as I'm concerned. Thank you for adding a workaround.

**Comment by fabiobaltieri** [comment]
Missed one test...

File: subsys/testsuite/ztest/include/zephyr/ztest.h [hunk 1]
```diff
@@ -28,6 +28,12 @@
 #error "You need to add CONFIG_ZTEST to your config file."
 #endif
 
+#if defined(CONFIG_NO_OPTIMIZATIONS) && !defined(CONFIG_ZTEST_ALLOW_NO_OPTIMIZATIONS)
+#error Running tests with CONFIG_NO_OPTIMIZATIONS is not supported and \
+	known to break many tests, please do not file issues about it. If you \
+	really mean to do this also set CONFIG_ZTEST_ALLOW_NO_OPTIMIZATIONS.
+#endif
+
```
Comment (yperess) [inline subsys/testsuite/ztest/include/zephyr/ztest.h:36]: Why can't we error in cmake? It seems that this is statically known.

**Review by yperess — CHANGES_REQUESTED** [review]

File: subsys/testsuite/ztest/include/zephyr/ztest.h (hunk 1 above)
Comment (fabiobaltieri) [reply subsys/testsuite/ztest/include/zephyr/ztest.h]: Grand idea, moved to a CMake message.

File: tests/bsim/bluetooth/audio/prj.conf [hunk 2]
```diff
@@ -3,6 +3,7 @@ CONFIG_BT_TESTING=y
 CONFIG_CONSOLE_HANDLER=y
 CONFIG_USERSPACE=n
 CONFIG_NO_OPTIMIZATIONS=y
+CONFIG_ZTEST_ALLOW_NO_OPTIMIZATIONS=y
```
Comment (Thalley) [inline tests/bsim/bluetooth/audio/prj.conf:6]: This is a babblesim test and is not related to ZTEST. I think you can remove this line :)

**Review by henrikbrixandersen — APPROVED** [review]

**Review by yperess — APPROVED** [review]

**Review by nashif — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

**Comment by aescolar** [comment]
@fabiobaltieri I don't think this change is a good idea as it got in.
Disabling optimizations is very normal both for debugging and getting coverage reports. Maybe a check of this kind  should be added as a check for tests which are known to not work without optimizations (maybe a build time error(?)) instead of a blanket/"default on" for all. 
I haven't been part of the talk but I presume the issue is due to code size and/or some being too slow otherwise?
It is a bit weird to on one hand say this https://github.com/zephyrproject-rtos/zephyr/blob/main/CMakeLists.txt#L709-L710 , and on the other hand enable it for all tests for a given board https://github.com/zephyrproject-rtos/zephyr/pull/58163/files .

**With this change coverage is broken in main**. You can just try, for ex.:
`twister -p native_posix -T tests/kernel/ --coverage`

CC @daor-oti @wopu-ot , note https://github.com/zephyrproject-rtos/zephyr/pull/58163

**Comment by fabiobaltieri** [comment]
> It is a bit weird to on one hand say this https://github.com/zephyrproject-rtos/zephyr/blob/main/CMakeLists.txt#L709-L710 , and on the other hand enable it for all tests for a given board https://github.com/zephyrproject-rtos/zephyr/pull/58163/files .

You are right that's contradictory, reworded in https://github.com/zephyrproject-rtos/zephyr/pull/58176.

Rest of the conversation in https://github.com/zephyrproject-rtos/zephyr/issues/58175.

**Pull request merged**